src/Stratis.Bitcoin/Utilities/MemoryCache.cs

```diff
@@ -0,0 +1,134 @@
+            this.keys = new LinkedList<CacheItem<TKey, TValue>>();
+            this.mutex = new object();
+        }
+
+        /// <summary>Gets the count of the current items for diagnostic purposes.</summary>
+        public int Count
+        {
+            get
+            {
+                lock (this.mutex)
+                {
+                    return this.keys.Count;
+                }
+            }
+        }
+
+        /// <summary>Create or overwrite an entry in the cache.</summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value to add to the cache.</param>
+        public void AddOrUpdate(TKey key, TValue value)
+        {
+            LinkedListNode<CacheItem<TKey, TValue>> node;
+
+            lock (this.mutex)
+            {
+                if (this.cache.TryGetValue(key, out node))
+                {
+                    node.Value.Value = value;
+                    this.keys.Remove(node);
+                }
+                else
+                {
+                    if (this.keys.Count == this.maxItemsCount)
+                    {
+                        // Remove 1 item.
+                        this.cache.Remove(this.keys.First.Value.Key);
+                        this.keys.RemoveFirst();
+                    }
+
+                    node = new LinkedListNode<CacheItem<TKey, TValue>>(new CacheItem<TKey, TValue>(key, value));
+                    this.cache.Add(key, node);
+                }
+
+                this.keys.AddLast(node);
+            }
+        }
+
+        /// <summary>Removes the object associated with the given key.</summary>
+        /// <param name="key">Item's key that will be removed from the cache.</param>
+        public void Remove(TKey key)
+        {
+            lock (this.mutex)
+            {
+                if (this.cache.TryGetValue(key, out LinkedListNode<CacheItem<TKey, TValue>> node))
+                {
+                    this.cache.Remove(node.Value.Key);
+                    this.keys.Remove(node);
+                }
+            }
+        }
+
+        /// <summary>Gets the item associated with this key if present.</summary>
+        /// <param name="key">Item's key.</param>
+        /// <param name="value">Item assosiated with specified <paramref name="key"/>.</param>
+        /// <returns><c>true</c> if cache contains the item; <c>false</c> otherwise.</returns>
+        public bool TryGetValue(TKey key, out TValue value)
+        {
+            lock (this.mutex)
+            {
+                if (this.cache.TryGetValue(key, out LinkedListNode<CacheItem<TKey, TValue>> node))
+                {
+                    this.keys.Remove(node);
+                    this.keys.AddLast(node);
+
+                    value = node.Value.Value;
+
+                    return true;
+                }
+            }
+
+            value = default(TValue);
+            return false;
+        }
+        
+        private class CacheItem<TKey, TValue>
```

Comment:
this should be the very first thing in the class, not the last
and pls document everything here

Reply:
Fixed

Reply:
Didn't comment the key,value properties because it's obvious what they represent and couldn't come up with anything better than "The key" and "The value"

Reply:
that should be ok